src/main/java/com/spikes2212/command/drivetrains/swerve/SwerveDrivetrain.java

```diff
@@ -0,0 +1,87 @@
+package com.spikes2212.command.drivetrains.swerve;
+
+import com.spikes2212.command.DashboardedSubsystem;
+import edu.wpi.first.math.geometry.Rotation2d;
+import edu.wpi.first.math.geometry.Translation2d;
+import edu.wpi.first.math.kinematics.ChassisSpeeds;
+import edu.wpi.first.math.kinematics.SwerveDriveKinematics;
+import edu.wpi.first.math.kinematics.SwerveModuleState;
+
+public abstract class SwerveDrivetrain extends DashboardedSubsystem {
+
+    private static final String NAMESPACE_NAME = "swerve drivetrain";
+
+    private final SwerveModule frontLeft;
+    private final SwerveModule frontRight;
+    private final SwerveModule backLeft;
+    private final SwerveModule backRight;
+
+    private final double trackWidth;
+    private final double trackLength;
+    private final double maxVelocity;
+    private final SwerveDriveKinematics kinematics;
+
+    private Rotation2d currentRobotAngle;
+
+    public SwerveDrivetrain(SwerveModule frontLeft, SwerveModule frontRight, SwerveModule backLeft, SwerveModule backRight, double trackWidth, double trackLength, double maxVelocity) {
+        super(NAMESPACE_NAME);
+        this.frontLeft = frontLeft;
+        this.frontRight = frontRight;
+        this.backLeft = backLeft;
+        this.backRight = backRight;
+        this.trackWidth = trackWidth;
+        this.trackLength = trackLength;
+        this.maxVelocity = maxVelocity;
+
+        Translation2d frontLeftWheelDistanceFromCenter = new Translation2d(trackWidth / 2, trackLength / 2);
+        Translation2d frontRightWheelDistanceFromCenter = new Translation2d(trackWidth / 2, -trackLength / 2);
+        Translation2d backLeftWheelDistanceFromCenter = new Translation2d(-trackWidth / 2, trackLength / 2);
+        Translation2d backRightWheelDistanceFromCenter = new Translation2d(-trackWidth / 2, -trackLength / 2);
+
+        kinematics = new SwerveDriveKinematics(frontLeftWheelDistanceFromCenter,
+                frontRightWheelDistanceFromCenter, backLeftWheelDistanceFromCenter,
+                backRightWheelDistanceFromCenter);
+    }
+
+    public abstract void setCurrentRobotAngle();
+
+    public void drive(double xSpeed, double ySpeed, double rotationSpeed, boolean isFieldRelative, double timeStep) {
+        SwerveModuleState[] states = kinematics.toSwerveModuleStates(getChassisSpeeds(isFieldRelative, xSpeed, ySpeed,
+                rotationSpeed, timeStep), new Translation2d());
```

Comment:
im pretty sure the default center of mass is (0, 0), so constructing a new `Translation2d` object is redundant